### ISSUE TITLE
Data loss when capturing from microphone eCaptureStatus::Stalled

### ISSUE THREAD

**Issue by emcodem** [Const-me/Whisper#145]
Dear Const, 
i spent the whole day to get out why some portions of live transcriptions are just missing and ended up dumping the raw audio in and out:

IN: ContextImpl.capture.cpp, Capture::readSample
OUT: ContextImpl.misc.cpp ContextImpl::runFull
Also i captured in paralell the actual microphone sound using ffmpeg. 

After comparing the 3 capture, it turns out that the missing portions come from Capture::readSample actually not receiving the corresponding samples (e.g. 1 second missing once every minute maybe)

Also, i added a log line in the section 
`if( hasStateFlag( eCaptureStatus::Stalled ) )`
and it seems that i come into the Stalled section from time to time. 

Unfortunately i have no idea about this obscure Media Foundation, so i started implementing read from stdin functionality in hope that i can outsource all audio related troubles to a 3rd party program e.g. ffmpeg and let whisper only do transcription and nothing else. 
However, it would be nice if you could give a hint how it can happen that we get into the stalled state?

On a sidenote i don't really have a congestion/bottleneck on hardware, it is only like 30% avg GPU utilisation when running a live whisper instance. 

Comments say:
		// We don't want concurrency here because it's not implemented, and will simply crash.
		// The "Stalled" flag which will cause capture to drop further samples.

Might a simple buffer help?

**Comment by Const-me** [comment]
@emcodem The stall state of that class should handle the case when the model doesn’t run fast enough to consume realtime audio. If I do nothing and keep accumulating audio in that buffer, the software will eventually consume all RAM, then fill all disk space with page file, then fail with out of memory runtime error. Dropping data is not ideal, but IMO still better outcome than the above horror.

I don’t know what specifically the task manager is measuring for that graph, but I would expect it to measure load of GPU shader cores. OTOH, I have a suspicion the compute shaders in this repository bottlenecks on VRAM bandwidth, not on GPU compute.

Assuming you use the C++ desktop app, one thing you could try, adjust these two numbers
https://github.com/Const-me/Whisper/blob/1.11.0/Examples/WhisperDesktop/CaptureDlg.cpp#L256-L257
Change these min/max numbers into 10/11 seconds, or 20/21 seconds.

It’s possible the relative difference between min/max causes the stall. One chunk used maximum 11 seconds, then next chunk used minimum 7 seconds, but 7 seconds wasn’t enough to transcribe the previous 11 seconds buffer.

As for that comment about concurrency, it’s technically possible to transcribe two chunks of audio in parallel. Not implemented because relatively hard to do, but also I’m not sure that going to help in the sense that will improve overall throughput of the transcribing by a substantial margin.

**Comment by emcodem** [comment]
Thanks for the fast response!
I work with MicrophoneCS currently, as far as i quickly saw, it does not accept duration params but i print them in ContextImpl::runCapture and i see they are set to 7.0 and 11.0 (didnt quickly find where actually these values come from. 

When saying above i have 30%, i did not mean this is the max value the graph shows, the graph shows always 100% when whisper is transcribing but then goes back to 0 when its only capturing. It only takes between  15~25% realtime to transcribe, that is why i say 30% avg.

However, if you say there is no buffer, the values 7/11 are not bad at all actually. Anyway, i will work with this but i believe there will be no way around some buffering at this stage, of course a very large buffer might not make a lot of sense when considering the live usecase where some user waits for the output actually, but on the other hand we can't fear about this tiny audio data filling up our memory. 
If we would really fear about too low heap memory issues, we would need to redesign how the main sample works because it actually reads the whole audio (probably hours long) file and converts it to mel which is being kept in the heap over the whole runtime. It could also just parse 30second junks from the file instead (in which case we would not need to distinguish between file and live mode from whisper perspective anymore)

I'll try adding some buffering and play with min/max values and let you know how it goes...

**Comment by Const-me** [comment]
@emcodem Well, if you see these 100% bursts only taking 30% of time, the performance is probably adequate.

Try to add following line of code
```
// If the transcribe hasn’t yet finished, allow the buffer to grow up to maxDuration length, before starting to drop the samples
if( newSamples < captureParams.maxDuration ) return S_OK;
```
To that place: https://github.com/Const-me/Whisper/blob/1.11.0/Whisper/Whisper/ContextImpl.capture.cpp#L274

That’s a small change easy to try, but with some luck (depending on why specifically it stalls) might help.

**Comment by emcodem** [comment]
Thanks a lot, i am currently giving this a shot. It will take some tests until i can come up with a final result because the conversation with you brought up multiple lights now.  
In the meantime, here a patch for microphoneCS example introducing min and max time as well as adding missing help entry for -c and -ld params.  

[miccs_min_and_max_capture_duration.patch](https://github.com/Const-me/Whisper/files/12114642/miccs_min_and_max_capture_duration.patch)

Again, as you did not apply any PR's or patches in the past, i guess you'll not add this one too which is totally fine with me. But in case my work with your project goes on, i'll not stop sending patch files (i use it also as kind of documentation for myself)  (but no PR's unless you prefer)

**Comment by emcodem** [comment]
OK so today i have not seen any stall anymore. There are multiple things to say about this:
First, initially with the default min and max 7/11 my GPU graph looked like this:
![whisperlive_7_11](https://github.com/Const-me/Whisper/assets/53784010/6e027152-4b75-4252-9f56-575591d2185d)

I thought that looks OK and as it usually takes maximum 1/3rd realtime for transcibing, 7 seconds min should be fine BUT
i saw that teamviewer can take up to 30% of the GPU (especially when ffplay was running and drawing a MEL spectogram on screen).

Second, i changed min and max to 3/10 which did not only deliver very satisfying and much faster results (thanks to the ingenious VAD you wrote Const), but also it seems to spread the work much better on the GPU: 
 
![whisperlive_3_10](https://github.com/Const-me/Whisper/assets/53784010/3d98c5f3-27b4-4c35-aa43-76ee731eab4c)

But of course, values like 3/10 is only possible using the line you posted above 
`if( newSamples < captureParams.maxDuration ) return S_OK;`

Otherwise it would potentially stop recording in case a 10 second block is currently transcribed and VAD shoots directly a 3 second segment as follow-up, right?

Thank you very much for your help on this Const, the results are really really satisfying i must say ^^

**Comment by Const-me** [comment]
@emcodem Thanks for the testing. Published another version 1.12 with that [adjustment](https://github.com/Const-me/Whisper/commit/b5d1001b6b26f31103a3d6bbd32c9d2f6239dc2a).
